src/reducers/player.js

```diff
@@ -1,35 +1,48 @@
-import {ActionTypes} from '../constants';
-const {INITIAL_STATE, TILE_SELECTED} = ActionTypes;
+import {ActionTypes, CellStates} from '../constants';
+const {INITIAL_STATE, RACK_TILE_SELECTED, END_TURN, BOARD_TILE_SELECTED} = ActionTypes;
 
 export default function player(state = {}, action) {
-  if (action.turn !== state.id && action.type !== INITIAL_STATE) {
+  if (action.player !== state.id || (action.turn !== state.id && action.type !== INITIAL_STATE)) {
     return state;
   }
 
   switch (action.type) {
-    case TILE_SELECTED:
-      if (state.selected == '') {
+    case RACK_TILE_SELECTED:
+      if (state.selected < 0) {
         return Object.assign({}, state, {
-          selected: action.tileId
+          selected: action.selected
         });
       } else {
-        var newletter, oldletter, newletters;
-        newletter = state.letters[parseInt(action.tileId.substring(2, 3))];
-        oldletter = state.letters[parseInt(state.selected.substring(2, 3))];
-        newletters = state.letters.slice();
-        newletters[parseInt(state.selected.substring(2, 3))] = newletter;
-        newletters[parseInt(action.tileId.substring(2, 3))] = oldletter;
+        let newletters = state.letters.slice();
+        let oldletter = newletters[state.selected];
+        newletters[state.selected] = newletters[action.selected];
+        newletters[action.selected] = oldletter;
         return Object.assign({}, state, {
           letters: newletters,
-          selected: ''
+          selected: -1
         });
       }
+    case BOARD_TILE_SELECTED:
+      if (state.selected > -1 && action.cellstate === CellStates.EMPTY) {
+        let newletters = state.letters.slice();
+        newletters.splice(state.selected, 1);
+        return Object.assign({}, state, {
+          letters: newletters,
+          selected: -1
+        });
+      } else {
+        return state;
+      }
+    case END_TURN:
+      return Object.assign({}, state, {
+        selected: -1
+      });
     case INITIAL_STATE:
```

Comment:
generally i like to keep INITIAL_STATE at the top because that defines the shape of that reducer.